From e2fda688fc169d75e022cec374ecf4ac61aa8f46 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Mon, 29 Oct 2018 12:10:41 +0100
Subject: [PATCH v3 1/1] dm: video: correctly set the cursor position

The terminal escape sequence ESC [ <x> ; <y> H is used to set the cursor
position. According to the ECMA 48 standard the upper left corner in the
escape sequences is [1, 1]. The video uclass uses [0, 0] as upper left
corner.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
v3:
	update unit test
v2
	remove trailing whitespace
---
 drivers/video/vidconsole-uclass.c | 8 ++++++++
 test/dm/video.c                   | 2 +-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/video/vidconsole-uclass.c b/drivers/video/vidconsole-uclass.c
index 1874887f2f3..db40a1396b2 100644
--- a/drivers/video/vidconsole-uclass.c
+++ b/drivers/video/vidconsole-uclass.c
@@ -272,6 +272,14 @@ static void vidconsole_escape_char(struct udevice *dev, char ch)
 		s++;    /* ; */
 		s = parsenum(s, &col);
 
+		/*
+		 * Video origin is [0, 0], terminal origin is [1, 1].
+		 */
+		if (row)
+			--row;
+		if (col)
+			--col;
+
 		set_cursor_position(priv, row, col);
 
 		break;
diff --git a/test/dm/video.c b/test/dm/video.c
index 7def338058e..dc20dad43b6 100644
--- a/test/dm/video.c
+++ b/test/dm/video.c
@@ -178,7 +178,7 @@ static int dm_test_video_ansi(struct unit_test_state *uts)
 
 	/* test set-cursor: [%d;%df */
 	vidconsole_put_string(con, "abc"ANSI_ESC"[2;2fab"ANSI_ESC"[4;4fcd");
-	ut_asserteq(142, compress_frame_buffer(dev));
+	ut_asserteq(143, compress_frame_buffer(dev));
 
 	/* test colors (30-37 fg color, 40-47 bg color) */
 	vidconsole_put_string(con, ANSI_ESC"[30;41mfoo"); /* black on red */
-- 
2.19.1

